build: exclude npm test directories on Windows

richardlau opened the pull request:
npm test directories are excluded on other platforms by
`tools/install.py`. Do the same on Windows.

Fixes: https://github.com/nodejs/node/issues/22901

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

richardlau commented:
For reference, this is the relevant bit of `tools/install.py` that excludes the `test` directories: https://github.com/nodejs/node/blob/058c5b81cdbabe8989a194ba5d388f4c230f4af6/tools/install.py#L84-L89

richardlau commented on vcbuild.bat:
```diff
@@ -332,7 +332,7 @@ copy /Y ..\README.md node-v%FULLVERSION%-win-%target_arch%\ > nul
 copy /Y ..\CHANGELOG.md node-v%FULLVERSION%-win-%target_arch%\ > nul
 if errorlevel 1 echo Cannot copy CHANGELOG.md && goto package_error
-robocopy /e ..\deps\npm node-v%FULLVERSION%-win-%target_arch%\node_modules\npm > nul
+robocopy ..\deps\npm node-v%FULLVERSION%-win-%target_arch%\node_modules\npm /e /xd test > nul
```
[robocopy documentation](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/robocopy) has the syntax as `robocopy <Source> <Destination> [<File>[ ...]] [<Options>]` so I've moved the existing option (`/e`) to the end of the command line in addition to the new option to exclude `test` directories.

richardlau commented:
cc @nodejs/build-files @nodejs/platform-windows @nodejs/npm

richardlau commented:
Also note this is going to be tricky to test as the regular CI doesn't build the installer/7z/zip packages.

joyeecheung commented:
There is a https://github.com/nodejs/build/blob/master/jenkins/xml/smoketest-release.xml in the build repo, I wonder what it's for...can we use it? Is it somewhere on Jenkins?

richardlau commented:
> 
> 
> There is a https://github.com/nodejs/build/blob/master/jenkins/xml/smoketest-release.xml in the build repo, I wonder what it's for...can we use it? Is it somewhere on Jenkins?

Not (as far as I can see) on https://ci.nodejs.org/view/All/. Maybe it's on ci-release (which I don't have access to)?

refack commented:
Any chance we can make `install.py` windows compatible instead?

> [nodejs/build:jenkins/xml/smoketest-release.xml@`master`](https://github.com/nodejs/build/blob/master/jenkins/xml/smoketest-release.xml)

Landed under the title `jenkins: add job config files for deleted jobs`... Which makes sense since the fragile bits in `ci-release` are WRT to firewalls, certificates, and the specific workers that are mapped to it.

I triggered a test build https://ci-release.nodejs.org/job/iojs+release/3779/

richardlau commented:
> 
> 
> Any chance we can make `install.py` windows compatible instead?
> 

It's certainly possible but it's not a one line change. We'd have to take into account the differences between the directory layouts on Windows vs the Unix-like platforms and things like symlinks not being used on Windows. Also even though I know it's not a supported platform I wonder if changing `install.py` for Windows would affect mingw/msys.

https://github.com/nodejs/node/issues/22901#issuecomment-422637283

richardlau commented:
> I triggered a test build https://ci-release.nodejs.org/job/iojs+release/3779/

Thanks. Could you, or someone else who has access the release ci report back on the results when the build completes?

refack commented:
No `/test/` in npm
[node-v11.0.0-rc.0-win-x64.zip](https://github.com/nodejs/node/files/2406779/node-v11.0.0-rc.0-win-x64.zip)
```
7z l node-v11.0.0-rc.0-win-x64.zip | grep "x64/node_modules/npm/[^/]*$"
2018-05-26 02:00:21 ....A         3258         1471  node-v11.0.0-rc.0-win-x64/node_modules/npm/.mailmap
2018-08-03 02:00:25 ....A          350          195  node-v11.0.0-rc.0-win-x64/node_modules/npm/.npmignore
2018-05-26 02:00:21 ....A         1228          529  node-v11.0.0-rc.0-win-x64/node_modules/npm/.travis.yml
2018-05-24 13:56:30 ....A          993          598  node-v11.0.0-rc.0-win-x64/node_modules/npm/appveyor.yml
2018-09-03 06:00:20 ....A        23475         9843  node-v11.0.0-rc.0-win-x64/node_modules/npm/AUTHORS
2018-09-03 06:00:20 D....            0            0  node-v11.0.0-rc.0-win-x64/node_modules/npm/bin
2018-09-03 06:00:20 ....A        61603        20601  node-v11.0.0-rc.0-win-x64/node_modules/npm/CHANGELOG.md
2018-05-26 02:00:21 D....            0            0  node-v11.0.0-rc.0-win-x64/node_modules/npm/changelogs
2018-05-24 13:56:30 ....A          554          337  node-v11.0.0-rc.0-win-x64/node_modules/npm/configure
2018-09-03 06:00:20 ....A         4566         1715  node-v11.0.0-rc.0-win-x64/node_modules/npm/CONTRIBUTING.md
2018-05-24 13:56:30 D....            0            0  node-v11.0.0-rc.0-win-x64/node_modules/npm/doc
2018-09-03 06:00:20 D....            0            0  node-v11.0.0-rc.0-win-x64/node_modules/npm/html
2018-09-03 06:00:20 D....            0            0  node-v11.0.0-rc.0-win-x64/node_modules/npm/lib
2018-05-24 13:56:30 ....A         9977         3524  node-v11.0.0-rc.0-win-x64/node_modules/npm/LICENSE
2018-05-24 13:56:30 ....A          159          123  node-v11.0.0-rc.0-win-x64/node_modules/npm/make.bat
2018-05-26 02:00:21 ....A         5802         1469  node-v11.0.0-rc.0-win-x64/node_modules/npm/Makefile
2018-05-24 13:56:30 D....            0            0  node-v11.0.0-rc.0-win-x64/node_modules/npm/man
2018-09-03 06:00:21 D....            0            0  node-v11.0.0-rc.0-win-x64/node_modules/npm/node_modules
2018-09-03 06:00:21 ....A         7671         2218  node-v11.0.0-rc.0-win-x64/node_modules/npm/package.json
2018-09-03 06:00:20 ....A         4871         2144  node-v11.0.0-rc.0-win-x64/node_modules/npm/README.md
2018-09-03 06:00:21 D....            0            0  node-v11.0.0-rc.0-win-x64/node_modules/npm/scripts
```

richardlau commented:
I realised on the way home that this change only affects the 7z/zip packages. Will need to look at the WiX files to work out if we can do the equivalent for the msi installer.

richardlau commented:
I think this is the bit that generates the npm files for WiX: https://github.com/nodejs/node/blob/9d71e6a6071134789a30471302d9735250dea96c/tools/msvs/msi/nodemsi.wixproj#L76

refack commented:
> I think this is the bit that generates the npm files for WiX:

I think it should work of the directory generated by the `:package` label, i.e. `node-v%FULLVERSION%-win-%target_arch%\node_modules\npm`

richardlau commented:
Yes, I think working off the directory created for `:package` will be simpler than the official way to exclude files/directories with `heat`which is to use XSLT (which I fear would scare off even more contributors from the Windows install build files). Would need some refactoring as currently you can skip `package` and go to `msi` in which case the directory would not be created.

addaleax commented:
> I realised on the way home that this change only affects the 7z/zip packages. Will need to look at the WiX files to work out if we can do the equivalent for the msi installer.

@richardlau Is this WIP? I’m adding the label but please remove it if I’m mistaken about that.

richardlau commented:
> 
> 
> > I realised on the way home that this change only affects the 7z/zip packages. Will need to look at the WiX files to work out if we can do the equivalent for the msi installer.
> 
> @richardlau Is this WIP? I’m adding the label but please remove it if I’m mistaken about that.

Yes, thanks. I don't think we should merge this unless the contents of the msi installer match the 7z/zip packages.

joaocgreis commented:
@richardlau thanks for working on this! The relevant upstream issue is https://github.com/wixtoolset/issues/issues/4023, so there is no easy way to exclude the tests with Heat. It should be easy to get the `npm` folder from the package folder instead of `deps`, but `vcbuild` needs to be adapted to have a common section for `package` and `msi` that runs this `robocopy` line.

richardlau commented:
Updated:
- split part of `:package` to a `:stage_package` step that is run if either `msi` or `package` is set
- replace references to `..\..\..\deps\npm` with `..\..\..\Release\node-v$(FullVersion)-win-$(Platform)\node_modules\npm` in `tools/msvs/msi/nodemsi.wixproj`. Note that the existing `:package` step always staged in `Release` but there should be no difference wrt `npm` for debug/release and x64/x86 configurations (i.e. even the debug configurations reference `Release` to pick up the staged `npm`).

I gave this a quick build/install locally and no npm and/or npm dependency `test` directories were installed.  If this looks okay, please could someone with access to the release CI do a test build for others to try?

danbev commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/17444/

joaocgreis commented:
Resumed CI: https://ci.nodejs.org/job/node-test-pull-request/17460/
Test build: https://nodejs.org/download/test/v11.0.0-test2018092789c37c2e64/ (https://ci-release.nodejs.org/job/iojs+release/3789/)

richardlau commented:
Resume CI: https://ci.nodejs.org/job/node-test-pull-request/17477/

richardlau commented:
> 
> 
> This should be semver-major. While I can't think of a good reason anything would break, we are removing files from the published packages and installer. This already has the necessary TSC reviews, unless @joyeecheung or @jasnell withdraw their approvals.

@jasnell I added a second commit for the msi installer after your approval; does this still look good to you?

The most recent resume CI is green, so I'd like to land this early next week to make the `semver-major` cutoff for v11.x (https://github.com/nodejs/node/issues/22180#issue-348473536).

cc @nodejs/tsc (as per https://github.com/nodejs/node/blob/master/COLLABORATOR_GUIDE.md#involving-the-tsc since this is defensively `semver-major`).

richardlau commented:
Landed in https://github.com/nodejs/node/commit/b3b3f53a339135903133842888467a0e6798acd0.
